data/src/main/java/com/teamagam/gimelgimel/data/layers/LayersLocalCacheData.java

```diff
@@ -0,0 +1,68 @@
+package com.teamagam.gimelgimel.data.layers;
+
+import android.content.Context;
+
+import com.teamagam.gimelgimel.data.config.Constants;
+import com.teamagam.gimelgimel.domain.layers.LayersLocalCache;
+import com.teamagam.gimelgimel.domain.messages.entity.contents.VectorLayer;
+
+import java.io.File;
+import java.net.URI;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import rx.Observable;
+
+@Singleton
+public class LayersLocalCacheData implements LayersLocalCache {
+
+    @Inject
+    Context mContext;
+
+    @Inject
+    FilesDownloader mFilesDownloader;
+
+    @Inject
+    public LayersLocalCacheData() {
+    }
+
+    @Override
+    public Observable<URI> cache(VectorLayer vectorLayer) {
+        return Observable.just(vectorLayer)
+                .map(this::downloadToCache);
+    }
+
+    @Override
+    public boolean isCached(VectorLayer vectorLayer) {
+        File file = new File(generateTargetURI(vectorLayer).getPath());
+        return file.exists();
+    }
+
+    @Override
+    public URI getCachedURI(VectorLayer vectorLayer) {
+        return generateTargetURI(vectorLayer);
+    }
+
+    private URI downloadToCache(VectorLayer vectorLayer) {
+        URI targetURI = generateTargetURI(vectorLayer);
+        mFilesDownloader.download(vectorLayer.getRemoteUrl(), targetURI);
+        return getCachedURI(vectorLayer);
+    }
+
+    private URI generateTargetURI(VectorLayer vectorLayer) {
+        File externalFilesDir = mContext.getExternalFilesDir(null);
+
+        String fullFilename = String.valueOf(externalFilesDir) +
```

Comment:
String.valueOf is redundant (I understood it from the interactor's logger, where you just used mVectorLayer as a string)